File: backend/app/services/auth_service.py
from sqlalchemy import select
from sqlalchemy.orm import Session

from app.config import settings
from app.models.user import User
from app.schemas.auth import TokenPair, UserCreate
from app.utils.security import create_token, hash_password, verify_password


class AuthService:
    @staticmethod
    def register(db: Session, payload: UserCreate) -> User:
        by_email = db.scalar(select(User).where(User.email == payload.email))
        if by_email:
            raise ValueError("User with provided email already exists.")

        if payload.phone:
            by_phone = db.scalar(select(User).where(User.phone == payload.phone))
            if by_phone:
                raise ValueError("User with provided phone already exists.")

        user = User(
            full_name=payload.full_name,
            email=payload.email,
            phone=payload.phone,
            password_hash=hash_password(payload.password),
            role=payload.role,
            status=payload.status,
            profile_image_url=payload.profile_image_url,
            bio=payload.bio,
            city=payload.city,
            preferred_language=payload.preferred_language,
        )
        db.add(user)
        db.commit()
        db.refresh(user)
        return user

    @staticmethod
    def login(db: Session, email: str, password: str) -> User:
        user = db.scalar(select(User).where(User.email == email))
        if not user or not verify_password(password, user.password_hash):
            raise ValueError("Invalid email or password.")
        return user

    @staticmethod
    def create_token_pair(user_id: int) -> TokenPair:
        subject = str(user_id)
        access = create_token(
            subject=subject,
            token_type="access",
            expires_minutes=settings.jwt_access_token_exp_minutes,
        )
        refresh = create_token(
            subject=subject,
            token_type="refresh",
            expires_minutes=settings.jwt_refresh_token_exp_minutes,
        )
        return TokenPair(access_token=access, refresh_token=refresh)

File: flutter_app/lib/services/auth_service.dart
import 'package:dio/dio.dart';
import '../models/models.dart';

class AuthService {
  final Dio _dio;
  AuthService(this._dio);

  Future<AuthToken> login({required String email, required String password}) async {
    final r = await _dio.post('/auth/login', data: {'email': email, 'password': password});
    return AuthToken.fromJson(r.data);
  }

  Future<AuthToken> register({
    required String fullName,
    required String email,
    String? phone,
    required String password,
    required String confirmPassword,
    String preferredLanguage = 'en',
  }) async {
    final r = await _dio.post('/auth/register', data: {
      'full_name': fullName,
      'email': email,
      'phone': phone,
      'password': password,
      'confirm_password': confirmPassword,
      'preferred_language': preferredLanguage,
    });
    return AuthToken.fromJson(r.data);
  }

  Future<AuthToken> refresh(String refreshToken) async {
    final r = await _dio.post('/auth/refresh', data: {'refresh_token': refreshToken});
    return AuthToken.fromJson(r.data);
  }

  Future<void> changePassword({required String current, required String newPwd}) async {
    await _dio.post('/auth/change-password', data: {'current_password': current, 'new_password': newPwd});
  }

  Future<void> forgotPassword(String email) async {
    await _dio.post('/auth/forgot-password', data: {'email': email});
  }
}
